pkg/rbac/parser_test.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rbac
+
+import (
+	"testing"
+
+	"github.com/onsi/gomega"
+	rbacv1 "k8s.io/api/rbac/v1"
+)
+
+//From: https://github.com/kubernetes-sigs/controller-tools/issues/612, but altered slightly (#3 & #4 made distinct)
+//- kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machinedeployments;machinedeployments/status;machinedeployments/finalizers,verbs=get;list;watch;create;update;patch;delete
+//- kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machinedeployments;machinedeployments/finalizers,verbs=get;list;watch;update;patch
+//- kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machinedeployments,verbs=get;list;watch
+//- kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machinedeployments,verbs=update;patch
+var rules = []*Rule{
+	{
+		Groups:    []string{"cluster.x-k8s.io"},
+		Resources: []string{"machinedeployments", "machinedeployments/status", "machinedeployments/finalizers"},
+		Verbs:     []string{"get", "list", "watch", "create", "update", "patch", "delete"},
+	},
+	{
+		Groups:    []string{"cluster.x-k8s.io"},
+		Resources: []string{"machinedeployments", "machinedeployments/finalizers"},
+		Verbs:     []string{"get", "list", "watch", "update", "patch"},
+	},
+	{
+		Groups:    []string{"cluster.x-k8s.io"},
+		Resources: []string{"machinedeployments"},
+		Verbs:     []string{"get", "list", "watch"},
+	},
+	{
+		Groups:    []string{"cluster.x-k8s.io"},
+		Resources: []string{"machinedeployments"},
+		Verbs:     []string{"update", "patch"},
+	},
+}
+
+var nrules = []*NormalizedRule{}
+
+func init() {
+	for _, rule := range rules {
+		nrules = append(nrules, rule.Normalize())
+	}
+}
+
+func Test_Subsume_Simple(t *testing.T) {
+	g := gomega.NewWithT(t)
+	// subsumes all others:
+	for _, nrule := range nrules {
+		g.Expect(nrules[0].Subsumes(nrule)).To(gomega.BeTrue())
+	}
+
+	g.Expect(nrules[1].Subsumes(nrules[2])).To(gomega.BeTrue())
+	g.Expect(nrules[1].Subsumes(nrules[3])).To(gomega.BeTrue())
+
+	// distinct
+	g.Expect(nrules[3].Subsumes(nrules[2])).To(gomega.BeFalse())
+	g.Expect(nrules[2].Subsumes(nrules[3])).To(gomega.BeFalse())
+}
+
+func Test_SubsumesIsReflexive(t *testing.T) {
+	g := gomega.NewWithT(t)
```

Comment:
use some kind of property testing for this and following test?

Reply:
What do you mean with property testing?

Reply:
See updated tests; in this case using the `testing/quick` package. We generate arbitrary `Rule`s and then make assertions about them.